File: control/bumperbot_ws/install/bumperbot_utils/local/lib/python3.10/dist-packages/bumperbot_utils/trajectory_plotter.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from nav_msgs.msg import Odometry, Path
from geometry_msgs.msg import PoseStamped 

class TrajectoryPlotter(Node):
    def __init__(self):
        super().__init__("trajectory_plotter")

        self.declare_parameter("odometry_topic", "bumperbot_controller/odom")
        self.odom_topic = self.get_parameter("odometry_topic")
        self.odom_subscriber = self.create_subscription(Odometry, self.odom_topic.value, self.odomCallback, 10)
        self.trajPublisher = self.create_publisher(Path, "bumperbot_controller/trajectory", 10)

        self.path_msgs_ = Path()

        self.get_logger().info('Trajectory node started')
    

    def odomCallback(self, msg: Odometry):

        self.path_msgs_.header.frame_id = msg.header.frame_id
        #PoseStamped message from the odometry message
        pose = PoseStamped()
        pose.header.frame_id = msg.header.frame_id
        pose.header.stamp = msg.header.stamp 
        pose.pose = msg.pose.pose

        self.path_msgs_.poses.append(pose)

        self.trajPublisher.publish(self.path_msgs_)

        self.get_logger().info('Published trajectory')

def main():
    rclpy.init()
    traj_plotter = TrajectoryPlotter()
    rclpy.spin(traj_plotter)

    traj_plotter.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: control/bumperbot_ws/src/bumperbot_utils/src/trajectory_plotter.cpp
#include <rclcpp/rclcpp.hpp>
#include <nav_msgs/msg/path.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>

using std::placeholders::_1;

class TrajectoryPlotter : public rclcpp::Node
{
    public:
        TrajectoryPlotter() : Node("trajectory_plotter")
        {
            odom_sub_ = create_subscription<nav_msgs::msg::Odometry>("/bumperbot_controller/odom",
                 10, std::bind(&TrajectoryPlotter::odom_callback, this, _1));
            path_pub_ = create_publisher<nav_msgs::msg::Path>("/bumperbot_controller/Path", 10);

        }
    private:
        rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
        rclcpp::Publisher<nav_msgs::msg::Path>::SharedPtr path_pub_;
        nav_msgs::msg::Path path_;

        void odom_callback(const nav_msgs::msg::Odometry & msg)
        {
            path_.header.frame_id = msg.header.frame_id;
            geometry_msgs::msg::PoseStamped pose;
            pose.header.frame_id = msg.header.frame_id;
            pose.header.stamp = msg.header.stamp;
            pose.pose = msg.pose.pose;

            path_.poses.push_back(pose);
            path_.header.stamp = msg.header.stamp;


            path_pub_->publish(path_);
        }

};

int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TrajectoryPlotter>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

